File: 卡码/代码随想录/1/08-贪心算法/Go/03-困难/LeetCode_0968.go
//968. 监控二叉树
package main

import . "greedy/types"

/*
0:无覆盖
1:有摄像头
2:有覆盖
 */
func minCameraCover(root *TreeNode) int {
	res:=0
	var traversal func(root *TreeNode) int
	traversal=func(root *TreeNode) int{
		if root==nil {
			return 2
		}
		left:=traversal(root.Left)
		right:=traversal(root.Right)
		if left==2&&right==2 {
			return 0
		}else if left==0||right==0 {
			res++
			return 1
		}else if left==1||right==1 {
			return 2
		}
		return -1
	}
	if traversal(root)==0 {
		res++
	}
	return res
}

File: 卡码/代码随想录/1/08-贪心算法/Js/03-困难/LeetCode_0968.js
//968. 监控二叉树
import TreeNode from "../TreeNode.js"

/*
0:无覆盖
1:有摄像头
2:有覆盖
*/
const minCameraCover = function(root) {
	let res=0;
	const traversal=root=>{
		if(!root) return 2;
		const left=traversal(root.left);
		const right=traversal(root.right);
		if(left==2&&right==2) return 0;
		else if(left==0||right==0){
			res++;
			return 1;
		}else if(left==1||right==1){
			return 2;
		}
	}
	if(traversal(root)==0) res++;
	return res;
};